### DISCUSSION TITLE
Password last reset date is not updated when changing admin password in BackOffice

### DISCUSSION THREAD

**Discussion by maxime-morel** [PrestaShop/PrestaShop#33790] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When updating an admin password from the Employee page, the `last_passwd_gen` date in `ps_employee`  is not updated.

### Expected behavior

The field should be updated regardless of how the password is generated.
It's currently updated only from the "password lost" fonction. For security reasons, we want the field to be updated as well when changing manually the password from the Employee page in the BO.

This will be useful for admin, security experts or module developers in order to apply a specific password policy (for exemple, in order to enforce password expiration in the future.

### Steps to reproduce

1. Go to the Employee section in BO
2. Change the Employee password (your own password or someone else password)
3. Check the `last_passwd_gen` in the database. The field is not updated.

### PrestaShop version(s) where the bug happened

1.5~8.1

**Comment by Hlavtox** [comment]
I am not sure it should be updated. For both customer and employee, this value is updated on creating them AND when changing the password via reset.

It's only use in PrestaShop is to check interval between resets. That's why the field is called "last password generation".

```
 } elseif ((strtotime($employee->last_passwd_gen . '+' . Configuration::get('PS_PASSWD_TIME_BACK') . ' minutes') - time()) > 0) {
   $this->errors[] = $this->trans('You can reset your password every %interval% minute(s) only. Please try again later.', ['%interval%' => Configuration::get('PS_PASSWD_TIME_BACK')], 'Admin.Login.Notification');
```

If this field is updated on password change from backoffice, the user who will want to legitimately reset his password will see that he just resetted the password, even though he didn't do it.

If you want to implement custom password policy, I think you should track last password change separately and react to it. 👍

Ping @kpodemski

**Comment by hibatallahAouadni** [comment]
Hello @maxime-profileo 

Thanks for your issue and your PR!
But as @Hlavtox mentioned above, I don't think that's the correct expected behavior.
Let's check with Product Managers, ping @PrestaShop/product-team can someone please confirm that the expected behavior provided by @maxime-profileo is correct or not? 🙏 

PS: thanks @Hlavtox for your quick reply 🚀 

Thanks!

**Comment by maxime-morel** [comment]
Hello @Hlavtox @hibatallahAouadni

I understand your point.
Knowing when the password was changed for the last time still remain something important imo.
We could maybe change my issue in a feature request instead of a bug, and I would suggest :
- Adding a new field `last_passwd_changed` (I would prefer using the existing field and adapt the error message, but I understand your point).
- Applying the same logic to Customer object

Regards,

**Comment by kpodemski** [comment]
To me, it sounds like a feature request, indeed. The current field helps prevent flooding password reminders.

**Comment by Hlavtox** [comment]
As a personal opinion, I would like @maxime-profileo to implement this as a module. I personally hate this practice, because the final result is people changing one number at the end of the password, or adding current year. 👍 But I will of course respect opinion of product team and other maintainers.

Anyway, it should not be that hard to do. :-)
- Add new field `last_passwd_changed` to employee table or to custom table in case of a module.
- Add a new config `PS_PASSWORD_VALIDITY` or something like this to employeee settings. Either to the core or by using a form modifier. `Employee password expiration time` - `Never`/`3 months`/`xxx`
- During page load, check the date and redirect to password change page ... which does not exist now ... new controller? Maybe not that easy to do. 😄

**Comment by MatShir** [comment]
@Hlavtox you consider changing and resetting the password 2 different actions from an employee perspective?

**Comment by Hlavtox** [comment]
@MatShir Yes, because:

- Password reset: open page, enter email, go to mailbox, click link, set new password.
- Password change: open page, enter old password, set new password.

**Comment by kpodemski** [comment]
Hello @maxime-profileo 

In my opinion, if you're motivated to implement this feature as a separate field in the database and it later would allow module developers to base on that in their solutions, then I think you could do it. Let's wait for @MatShir approval.

**Comment by Hlavtox** [comment]
@kpodemski Why would we need a field in a database when this field would not be used by core at all? If a developer wants to implement this feature, adding a new table to store it is easy. 👍

**Comment by kpodemski** [comment]
@Hlavtox it would be use because it would react to manual change of the password, right?

but! to be totally clear, I don't see any blocking point for having it as a module

field could be in the other table and there are proper hooks for ObjectModel manipulation that should allow to do that

@maxime-profileo do you see anything other that would prevent you from doing that from the module?

**Comment by maxime-morel** [comment]
Hello @kpodemski 

Concerning the module, I'm not sure to see a hook that can be used for that.
In FO we have `actionPasswordRenew` but I don't see a similar hook the the BO.

Please note however that a module will not be able to retrieve historical data before it was installed. It's for this reason we were hoping to see this basic security feature as a native fonction.
We are generally requested, after a hacking/attempt, to check  what's the status of employees passwords. So this would require the merchant to already have the module in place (which will not be the case).

**Comment by kpodemski** [comment]
Hi @maxime-profileo 

> We are generally requested, after a hacking/attempt, to check what's the status of employees passwords. So this would require the merchant to already have the module in place (which will not be the case).

Do hooks like `actionObjectEmployeeUpdateBefore`, and `actionObjectEmployeeUpdateAfter` don't work?

In the end, it's a feature request, so I suggest closing your PR for now and thinking about the way that would allow you to create a module. As @Hlavtox said, having the additional field in the core doesn't do anything unless you'd like to implement a feature that requires employees to change passwords every month, for example 😎

**Comment by RosaBenouamer** [comment]
What about change the name of last_passwd_gen by last_passwd and add a flag precising the source of the modification. It avoid to add as many fields as there are ways to update ?

**Comment by RosaBenouamer** [comment]
@kpodemski ?

**Comment by RosaBenouamer** [comment]
As this is not an issue but a new need. I propose to convert to discussion
